File: DSA_python/DSA11018 XÂY DỰNG LẠI CÂY NHỊ PHÂN TÌM KIẾM.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def add_to_tree(root, node):
    if root.val > node:
        if not root.left: root.left = TreeNode(node)
        else: add_to_tree(root.left, node)
    elif root.val < node:
        if not root.right: root.right = TreeNode(node)
        else: add_to_tree(root.right, node)

def pre_order(node):
    if not node:
        return
    
    print(node.val, end=" ")
    pre_order(node.left)
    pre_order(node.right)

def submit():
    n = int(input())
    edges = list(map(int, input().split()))

    root = TreeNode(edges[0])
    for i in range(1, n):
        add_to_tree(root, edges[i])

    pre_order(root)
    print()











test = int(input())
for i in range(test):
    submit()

File: DSA/DSA11018 XÂY DỰNG LẠI CÂY NHỊ PHÂN TÌM KIẾM.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};


void add(TreeNode* node, int num)
{
    if (num < node->val)
    {
        if (node->left) add(node->left, num);
        else node->left = new TreeNode(num);
    }
    else if (num > node->val)
    {
        if (node->right) add(node->right, num);
        else node->right = new TreeNode(num);
    }
}

void post_order_traversal(TreeNode* node)
{
    if (!node)
    {
        return;
    }
    
    cout << node->val << " ";
    post_order_traversal(node->left);
    post_order_traversal(node->right);
}

void submit()
{
    int n;
    cin >> n;
    vector<int> node(n);
    for (auto &&i : node) cin >> i;
    
    TreeNode* root = new TreeNode(node[0]);

    for (int i = 1; i < n; i++)
    {
        add(root, node[i]);
    }
    
    post_order_traversal(root);

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
} 